### PR TITLE
Copy image

### LINKED ISSUE #53: Copy to clipboard feature
Pantheon Photos is already an amazing piece of software. Congratulations and thanks for your efforts!

However, what I very much miss is a copy to clipboard feature. 

Quite often I want to use a photo and upload it to imgur or attach it to an email. Currently I always have to show the photo in the file manager to be able to get it.

Is it possible to implement a copy+c feature by which I can copy the selected photo to the clipboard and can quickly use it in another program?

Thanks for listening!

Launchpad Details: [#LP1309146](https://bugs.launchpad.net/bugs/1309146) Alba Nader - 2014-04-17 19:10:29 +0000

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/61472729-copy-to-clipboard-feature?utm_campaign=plugin&utm_content=tracker%2F60550306&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60550306&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jeremypw**
Fixes #53 

Replaces #568 

This PR extends #568 to copy several different formats onto the clipboard.  This allows pasting the requested data type into different apps.  Further enhancement includes copying several image files from the CollectionView and adding another option to copy the image metadata as text instead of the uri.

Note that pasting images into Github comments works but only for one file at a time for some reason (pasting multiple files only transfers the first). 

Also pasting into Inkscape only works for single images as it is not clear how to store multiple image data into the clipboard.

However pasting multiple files into Files does work.

### CODE REVIEW

**Comment by TheGag96** [comment]
Hey, thanks for doing this! I've moved on from Elementary but I'm glad to see this feature in there.

**Review by zeebok — COMMENTED** [review]
This seems good. I do not *quite* get the difference between copying metadata vs the image (or maybe just the use case) but my concern is it being unclear when to use which and what file types are supported by them.

**Review by zeebok — APPROVED** [review]

**Comment by jeremypw** [comment]
@zeebok Thanks for looking at this!  I need to refresh my memory about the details then I'll try to answer your questions.

File: src/Views/EditingHostPage.vala [hunk 1]
```diff
@@ -28,6 +28,24 @@ public abstract class EditingHostPage : SinglePhotoPage {
...
 
+    private enum TargetType {
+        GNOME_COPIED_FILES,
+        IMAGE,
+        TEXT_URI_LIST,
+        TEXT
+    }
+
+    private const Gtk.TargetEntry [] COPY_TARGETS = {
+        {"image/png", Gtk.TargetFlags.OTHER_APP, TargetType.IMAGE}, // First parameter name is critical
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply src/Views/EditingHostPage.vala:39]: This is to ensure it works with any other app that *accepts* images in `image/png` format (which should be most non-proprietary apps at least) in which case we can put Pixbuf data into the clipboard. As there is a general function for the Photo class to obtain a Pixbuf from them it *should* work with most source formats that Photos can handle.

**Comment by jeremypw** [comment]
@zeebok I have answered your query re `image/png` target inline.  Regarding "Copy Metadata" this is to give the option of copying the information about the photo shown in the Info Sidebar into an app that accepts text.  If you just copy a photo into such a document then it is the URI to the photo file that is transferred.   As to use cases the main one I can think of the top of my head is when reporting bugs about image handling it can be useful to include this information.   It might be better put this action into the sidebar itself on reflection but as I was adding actions to the context menu in this PR it was easier to put it there.

File: src/Views/EditingHostPage.vala (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeebok) [reply src/Views/EditingHostPage.vala:39]: Interesting, thanks. Not super familiar with how things like this works so I appreciate the explanation

**Pull request merged**